core/types/celo_receipt_test.go

```diff
@@ -101,3 +102,119 @@ func checkStorageRLPEncodeDecodeConsistency(r *ReceiptForStorage, t *testing.T)
 }
+
+// Tests that the effective gas price is correctly derived for different transaction types, in different scenarios.
+func TestReceiptEffectiveGasPriceDerivation(t *testing.T) {
+	gasPrice := big.NewInt(1000)
+	gasFeeCap := big.NewInt(800)
+	gasTipCap := big.NewInt(100)
+	config := params.TestChainConfig
+	config.GingerbreadBlock = big.NewInt(1)
+	// Receipt base fee is the base fee encoded in the receipt which will be set post cel2 for CeloDynamicFeeTxV2 types.
+	receiptBaseFee := big.NewInt(50)
+
+	t.Run("LegacyTx", func(t *testing.T) {
+		testNonDynamic(t, NewTransaction(0, common.Address{}, big.NewInt(0), 0, gasPrice, nil), gasPrice)
+	})
+	t.Run("AccessListTx", func(t *testing.T) {
+		testNonDynamic(t, NewTx(&AccessListTx{GasPrice: gasPrice}), gasPrice)
+	})
+	t.Run("DynamicFeeTx", func(t *testing.T) {
+		tx := NewTx(&DynamicFeeTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+	})
+	t.Run("CeloDynamicFeeTx", func(t *testing.T) {
+		tx := NewTx(&CeloDynamicFeeTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+		tx = NewTx(&CeloDynamicFeeTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap, FeeCurrency: &common.Address{}})
+		testDynamicWithFeeCurrency(t, tx, nil)
+	})
+	t.Run("CeloDynamicFeeTxV2", func(t *testing.T) {
+		tx := NewTx(&CeloDynamicFeeTxV2{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+		testDynamic(t, tx, receiptBaseFee)
+		tx = NewTx(&CeloDynamicFeeTxV2{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap, FeeCurrency: &common.Address{}})
+		testDynamicWithFeeCurrency(t, tx, nil)
+		testDynamicWithFeeCurrency(t, tx, receiptBaseFee)
+	})
+	t.Run("CeloDenominatedTx", func(t *testing.T) {
+		tx := NewTx(&CeloDenominatedTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+		tx = NewTx(&CeloDynamicFeeTxV2{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap, FeeCurrency: &common.Address{}})
+		testDynamicWithFeeCurrency(t, tx, nil)
+	})
+}
+
+func testNonDynamic(t *testing.T, tx *Transaction, receiptBaseFee *big.Int) {
+	// Non dynamic txs should always have the gas price defined in the tx.
+	config := params.TestChainConfig
+	config.GingerbreadBlock = big.NewInt(1)
+	preGingerbreadBlock := uint64(0)
+	postGingerbreadBlock := uint64(2)
+
+	receipts := []*Receipt{{BaseFee: receiptBaseFee}}
+	txs := []*Transaction{tx}
+
+	// Pre-gingerbread
+	err := Receipts(receipts).DeriveFields(config, blockHash, preGingerbreadBlock, blockTime, nil, nil, txs)
+	require.NoError(t, err)
+	require.Equal(t, tx.GasPrice(), receipts[0].EffectiveGasPrice)
+
+	// Post-gingerbread
+	err = Receipts(receipts).DeriveFields(config, blockHash, postGingerbreadBlock, blockTime, baseFee, nil, txs)
+	require.NoError(t, err)
+	require.Equal(t, tx.GasPrice(), receipts[0].EffectiveGasPrice)
+}
+
+func testDynamic(t *testing.T, tx *Transaction, receiptBaseFee *big.Int) {
+	// Dynamic txs with no fee currency shold have nil for the effective gas price pre-gingerbread and the effective gas
```

Comment:
nit: Why not make those docstrings for the function?

Reply:
No reason, done.